python_modules/dagster/dagster/_daemon/asset_daemon.py

```diff
@@ -226,51 +327,266 @@ def _run_iteration_impl(
+                        max_workers=settings.get("num_workers"),
+                        thread_name_prefix="asset_daemon_worker",
+                    )
+                )
+
+            while True:
+                start_time = pendulum.now("UTC").timestamp()
+                yield from self._run_iteration_impl(
+                    workspace_process_context,
+                    threadpool_executor=threadpool_executor,
+                    amp_tick_futures=amp_tick_futures,
+                    last_submit_times=last_submit_times,
+                    debug_crash_flags={},
+                )
+                yield None
+                end_time = pendulum.now("UTC").timestamp()
+                loop_duration = end_time - start_time
+                sleep_time = max(0, MIN_INTERVAL_LOOP_TIME - loop_duration)
+                shutdown_event.wait(sleep_time)
+                yield None
+
+    def _run_iteration_impl(
+        self,
+        workspace_process_context: IWorkspaceProcessContext,
+        threadpool_executor: Optional[ThreadPoolExecutor],
+        amp_tick_futures: Dict[Optional[str], Future],
+        last_submit_times: Dict[Optional[str], float],
+        debug_crash_flags: SingleInstigatorDebugCrashFlags,  # TODO No longer single instigator
+    ):
+        workspace = workspace_process_context.create_request_context()
+        asset_graph = ExternalAssetGraph.from_workspace(workspace)
+
+        instance: DagsterInstance = workspace_process_context.instance
+        settings = instance.get_settings("auto_materialize")
+
+        now = time.time()
+
+        if get_auto_materialize_paused(instance):
+            yield
+            return
+
+        if not self._initialized_evaluation_id:
+            self._initialize_evaluation_id(instance, asset_graph)
+            self._initialized_evaluation_id = True
+
+        group_origins: Set[Optional[ExternalInstigatorOrigin]]
+
+        if settings.get("use_asset_policy_sensors"):
+            # TODO This will include user-defined evaluation groups once there is a
+            # user-facing way to set them. For now, every repository with at least one
+            # asset with an AMP policy or auto-observe interval gets a group for that repository
+
+            group_origins = set()
+
+            for asset_key, policy in asset_graph.auto_materialize_policies_by_key.items():
+                if not policy:
+                    continue
+                origin = get_instigator_origin_for_asset_key(asset_key, asset_graph)
+                if origin:
+                    group_origins.add(origin)
+
+            for asset_key in asset_graph.source_asset_keys:
+                if asset_graph.get_auto_observe_interval_minutes(asset_key) is not None:
+                    continue
+
+                origin = get_instigator_origin_for_asset_key(asset_key, asset_graph)
+                if origin:
+                    group_origins.add(origin)
```

Comment:
@sryza this PR is quite large but gives you a sneak peek at where your code would plug in - we would just need a way for, for each asset key in the asset graph, a way to map it onto a (sensor name, repo name, code location name) tuple (which is all an 'instigator origin' is)